src/handlers.rs

```diff
@@ -279,4 +396,101 @@ mod test {
             handle_request(&test_logger(), &request).expect("should handle request with no error");
         assert_eq!(expected, output);
     }
+
+    #[test]
+    fn test_handle_request_getai() {
+        let request = protocol::Request {
+            ty: protocol::RequestType::GETAI,
+            key: &CString::new("localhost".to_string())
+                .unwrap()
+                .into_bytes_with_nul(),
+        };
+
+        // The getaddrinfo addresses ordering is quite unpredictable.
+        // Trying the two potential orderings.
+        // We could also be on a IPv4-only machine on CI.
```

Comment:
Is this because `/etc/hosts` doesn't contain the entry, or because the loopback interface in CI doesn't contain a v6 address?

I fear this might also be a problem on other machines running `cargo test` - we should make this a bit more reliable…

Reply:
> Is this because /etc/hosts doesn't contain the entry, or because the loopback interface in CI doesn't contain a v6 address?

It's an issue on kernels configured not to support IPv6.

> I fear this might also be a problem on other machines running cargo test - we should make this a bit more reliable…

Yeah, I'm not super happy with that either.

The idea here is not only to test the serialized output itself, but also the content of the response. What we're sending to the wire is not exactly what `getaddrinfo` returns. There's some mapping logic happening in the `handlers::getaddrinfo` function, that's partially what we're testing here.

I put some thoughts on that and did not find a more robust way to check this function output than resolving `localhost` :( 

Do you have any idea?

Reply:
I think:
 - we should test the `serialize_` with manually crafted input
 - test a `GETAI` request for `localhost` to not fail (but don't peek into the response)
 - … and expect `dns_lookup::getaddrinfo` to do the right thing.

Reply:
> we should test the serialize_ with manually crafted input

This is already tested by https://github.com/twosigma/nsncd/pull/39/files#diff-148e623ad4cdaab49b81901460099f4efe7900e8c57e4f7a0ff6778589be01f8R464

> test a GETAI request for localhost to not fail (but don't peek into the response) … and expect dns_lookup::getaddrinfo to do the right thing.

This test is not really testing `dns_lookup::getaddrinfo` but rather than `handlers::getaddrinfo` properly massage its result to our needs. Removing the duplicates entries for now. getaddrinfo returns one item per `AddrInfo::socktype`, we only care about the IP adresses, we want to filter out the duplicates.

Reply:
WDYT about https://github.com/flokli/nsncd/commit/d7f89f81cdb6e988fdf42b2479907e65b3ca459c?

I still don't know how to handle the fact that localhost might only return v4 addresses, depending on the host configured, but at least now this tests the deduplication again.

Reply:
sounds good